File: src/main/java/miniDB/index/bptree/TreeInternalNode.java
package miniDB.index.bptree;

import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.LinkedList;


/**
 *
 * Class for our Internal nodes
 *
 */
@SuppressWarnings({"WeakerAccess", "unused"})
class TreeInternalNode extends TreeNode {

    private final LinkedList<Long> pointerArray;  // the pointer array

    /**
     * Create an internal node
     *
     * @param nodeType the node type parameter
     * @param pageIndex the index of the page
     */
    TreeInternalNode(TreeNodeType nodeType, long pageIndex) {
        super(nodeType, pageIndex);
        pointerArray = new LinkedList<>();
    }

    void removePointerAt(int index)
        {pointerArray.remove(index);}

    long getPointerAt(int index) {
        return((index < 0 || index >= pointerArray.size()) ? -1 : pointerArray.get(index));}

    long popPointer()
        {return(pointerArray.pop());}

    long removeLastPointer()
        {return(pointerArray.removeLast());}

    void addPointerAt(int index, long val)
        {pointerArray.add(index, val);}

    void addPointerLast(long val)
        {pointerArray.addLast(val);}

    void setPointerAt(int index, long val)
        {pointerArray.set(index, val);}

    int getPointerListSize()
        {return(pointerArray.size());}

    void pushToPointerArray(long val)
        {pointerArray.push(val);}


    /**
     * @param r pointer to *opened* B+ tree file
     * @throws IOException is thrown when an I/O exception is captured.
     */
    @Override
    public void writeNode(RandomAccessFile r, BPlusConfiguration conf)
            throws IOException {

        // update root index in the file
        if(this.isRoot()) {
            r.seek(conf.headerSize-16L);
            r.writeLong(getPageIndex());
        }

        // account for the header page as well.
        r.seek(getPageIndex());

        byte[] buffer = new byte[conf.pageSize];
        ByteBuffer bbuffer = ByteBuffer.wrap(buffer);bbuffer.order(ByteOrder.BIG_ENDIAN);
        // write the node type
        bbuffer.putShort(getPageType());

        // write current capacity
        bbuffer.putInt(getCurrentCapacity());

        // now write Key/Pointer pairs
        for(int i = 0; i < getCurrentCapacity(); i++) {
            bbuffer.putLong(getPointerAt(i));   // Pointer
            conf.writeKey(bbuffer, getKeyAt(i));
        }
        // final pointer.
        bbuffer.putLong(getPointerAt(getCurrentCapacity()));
        r.write(buffer);
    }
}

File: src/main/kotlin/com/lss233/minidb/engine/index/bptree/TreeInternalNode.kt
package com.lss233.minidb.engine.index.bptree

import java.io.IOException
import java.io.RandomAccessFile
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*

/**
 *
 * Class for our Internal nodes
 *
 * Create an internal node
 *
 * @param nodeType the node type parameter
 * @param pageIndex the index of the page
 */
@Suppress("unused")
class TreeInternalNode(nodeType: TreeNodeType, pageIndex: Long):TreeNode(nodeType, pageIndex) {

    private val pointerArray: LinkedList<Long> = LinkedList()

    fun removePointerAt(index: Int) {
        pointerArray.removeAt(index)
    }

    fun getPointerAt(index: Int): Long {
        return if (index < 0 || index >= pointerArray.size) -1 else pointerArray[index]
    }

    fun popPointer(): Long {
        return pointerArray.pop()
    }

    fun removeLastPointer(): Long {
        return pointerArray.removeLast()
    }

    fun addPointerAt(index: Int, `val`: Long) {
        pointerArray.add(index, `val`)
    }

    fun addPointerLast(`val`: Long) {
        pointerArray.addLast(`val`)
    }

    fun setPointerAt(index: Int, `val`: Long) {
        pointerArray[index] = `val`
    }

    val pointerListSize: Int
        get() = pointerArray.size

    fun pushToPointerArray(`val`: Long) {
        pointerArray.push(`val`)
    }

    /**
     * @param r pointer to *opened* B+ tree file
     * @throws IOException is thrown when an I/O exception is captured.
     */
    @Throws(IOException::class)
    override fun writeNode(r: RandomAccessFile, conf: BPlusConfiguration) {

        // update root index in the file
        if (isRoot()) {
            r.seek(conf.headerSize - 16L)
            r.writeLong(pageIndex)
        }

        // account for the header page as well.
        r.seek(pageIndex)
        val buffer = ByteArray(conf.pageSize)
        val bbuffer = ByteBuffer.wrap(buffer)
        bbuffer.order(ByteOrder.BIG_ENDIAN)
        // write the node type
        bbuffer.putShort(getPageType())

        // write current capacity
        bbuffer.putInt(getCurrentCapacity())

        // now write Key/Pointer pairs
        for (i in 0 until getCurrentCapacity()) {
            bbuffer.putLong(getPointerAt(i)) // Pointer
            conf.writeKey(bbuffer, getKeyAt(i))
        }
        // final pointer.
        bbuffer.putLong(getPointerAt(getCurrentCapacity()))
        r.write(buffer)
    }
}
